### PR TITLE
implement handling of missing attributes

### PR DESCRIPTION
**Pull request by mbauerness**
- Throws an Exception when a required attribute is missing
- Also fixes the cobertura bug where branches were calculated wrong

### CODE REVIEW

File: src/main/java/edu/hm/hafner/metric/parser/XmlParser.java [hunk 1]
```diff
@@ -31,6 +32,9 @@ protected void setRootNode(final ModuleNode newRootNode) {
         rootNode = newRootNode;
     }
 
+    private static final String OPTIONAL_JACOCO_ATTRIBUTE = "line";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/metric/parser/XmlParser.java:35]: Move that to the corresponding parser

File: src/main/java/edu/hm/hafner/metric/parser/XmlParser.java [hunk 2]
```diff
@@ -81,6 +85,32 @@ public ModuleNode parse(final Reader reader) {
...
+
+        return value.getValue();
+    }
+
+    /**
+     * Checks if the given attribute is an optional one in the report or not.
+     * @param attribute the attribute of the report
+     * @return if the attribute is optional
+     */
+    private boolean isOptional(final String attribute) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/metric/parser/XmlParser.java:106]: This should be an abstract method that will be overridden in the concrete classes. (Or non abstract with a default implementation).

File: src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java [hunk 3]
```diff
@@ -170,6 +172,77 @@ private void verifyCoverageMetrics(final Node tree) {
...
+    @Test @Disabled("How should we handle these different results?")
+    void shouldCreateTreeOfAnalysisModel() {
+        Node tree;
+        try (FileInputStream stream = new FileInputStream(
+                "src/test/resources/jacoco-analysis-model.xml"); InputStreamReader reader = new InputStreamReader(stream)) {
+            tree = new JacocoParser().parse(reader);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java:184]: Extract the filename and reuse the existing code.

File: src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java [hunk 4]
```diff
@@ -170,6 +172,77 @@ private void verifyCoverageMetrics(final Node tree) {
         assertThat(tree).hasName(PROJECT_NAME).doesNotHaveParent().isRoot().hasMetric(MODULE).hasParentName("^");
     }
 
+    @Test @Disabled("How should we handle these different results?")
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java:175]: Is this about missing attributes or about the other problem with values on different levels?

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java (hunk 4 above)
Comment (mbauerness) [reply src/test/java/edu/hm/hafner/metric/parser/JacocoParserTest.java:175]: It's the test case for the file where we count more classes/instructions than JaCoCo

**Review by uhafner — APPROVED** [review]

**Pull request merged**